kv/kvserver: TestReplicateQueueUpAndDownReplicateNonVoters failed

kv/kvserver.TestReplicateQueueUpAndDownReplicateNonVoters [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4150396&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4150396&tab=artifacts#/) on master @ [912964e02ddd951c77d4f71981ae18b3894e9084](https://github.com/cockroachdb/cockroach/commits/912964e02ddd951c77d4f71981ae18b3894e9084):


```
=== RUN   TestReplicateQueueUpAndDownReplicateNonVoters
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicateQueueUpAndDownReplicateNonVoters2912775434
    test_log_scope.go:80: use -show-logs to present logs inline
    replicate_queue_test.go:368: condition failed to evaluate within 45s: expected upreplication to 2 non-voters; found 0
        goroutine 95234450 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin({0x53a38d8, 0xc00dd0a340}, 0x0, 0xc015ece6c0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60 +0x5f
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon({0x53a38d8, 0xc00dd0a340}, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41 +0x4a
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicateQueueUpAndDownReplicateNonVoters(0xc00dd0a340)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replicate_queue_test.go:368 +0x4e5
        testing.tRunner(0xc00dd0a340, 0x43ec660)
        	/usr/local/go/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1306 +0x35a
    panic.go:642: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicateQueueUpAndDownReplicateNonVoters2912775434
--- FAIL: TestReplicateQueueUpAndDownReplicateNonVoters (46.41s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=2
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateQueueUpAndDownReplicateNonVoters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -301,23 +301,22 @@ func TestReplicateQueueDownReplicate(t *testing.T) {
 }
 
 func scanAndGetNumNonVoters(
-	ctx context.Context,
-	t *testing.T,
-	tc *testcluster.TestCluster,
-	store *kvserver.Store,
-	scratchKey roachpb.Key,
+	t *testing.T, tc *testcluster.TestCluster, scratchKey roachpb.Key,
 ) (numNonVoters int) {
-	// Nudge the replicateQueue to up/down-replicate our scratch range.
-	if err := store.ForceReplicationScanAndProcess(); err != nil {
-		t.Fatal(err)
+	for _, s := range tc.Servers {
+		// Nudge internal queues to up/down-replicate our scratch range.
+		require.NoError(t, s.Stores().VisitStores(func(s *kvserver.Store) error {
+			require.NoError(t, s.ForceSplitScanAndProcess())
+			require.NoError(t, s.ForceReplicationScanAndProcess())
+			require.NoError(t, s.ForceRaftSnapshotQueueProcess())
+			return nil
+		}))
 	}
 	scratchRange := tc.LookupRangeOrFatal(t, scratchKey)
 	row := tc.ServerConn(0).QueryRow(
-		`SELECT array_length(non_voting_replicas, 1) FROM crdb_internal.ranges_no_leases WHERE range_id=$1`,
+		`SELECT coalesce(max(array_length(non_voting_replicas, 1)),0) FROM crdb_internal.ranges_no_leases WHERE range_id=$1`,
 		scratchRange.GetRangeID())
-	err := row.Scan(&numNonVoters)
-	log.Warningf(ctx, "error while retrieving the number of non-voters: %s", err)
-
+	require.NoError(t, row.Scan(&numNonVoters))
 	return numNonVoters
 }
 
@@ -329,7 +328,6 @@ func TestReplicateQueueUpAndDownReplicateNonVoters(t *testing.T) {
 	skip.UnderRace(t)
 	defer log.Scope(t).Close(t)
 
-	ctx := context.Background()
 	tc := testcluster.StartTestCluster(t, 1,
 		base.TestClusterArgs{
 			ReplicationMode: base.ReplicationAuto,
@@ -360,13 +358,10 @@ func TestReplicateQueueUpAndDownReplicateNonVoters(t *testing.T) {
 	// Add two new servers and expect that 2 non-voters are added to the range.
 	tc.AddAndStartServer(t, base.TestServerArgs{})
 	tc.AddAndStartServer(t, base.TestServerArgs{})
-	store, err := tc.Server(0).GetStores().(*kvserver.Stores).GetStore(
-		tc.Server(0).GetFirstStoreID())
-	require.NoError(t, err)
 
 	var expectedNonVoterCount = 2
 	testutils.SucceedsSoon(t, func() error {
-		if found := scanAndGetNumNonVoters(ctx, t, tc, store, scratchKey); found != expectedNonVoterCount {
+		if found := scanAndGetNumNonVoters(t, tc, scratchKey); found != expectedNonVoterCount {
 			return errors.Errorf("expected upreplication to %d non-voters; found %d",
 				expectedNonVoterCount, found)
 		}
@@ -379,7 +374,7 @@ func TestReplicateQueueUpAndDownReplicateNonVoters(t *testing.T) {
 	require.NoError(t, err)
 	expectedNonVoterCount = 0
 	testutils.SucceedsSoon(t, func() error {
-		if found := scanAndGetNumNonVoters(ctx, t, tc, store, scratchKey); found != expectedNonVoterCount {
+		if found := scanAndGetNumNonVoters(t, tc, scratchKey); found != expectedNonVoterCount {
 			return errors.Errorf("expected downreplication to %d non-voters; found %d",
 				expectedNonVoterCount, found)
 		}

